File: seung365/programmers/18.py
def hash (arr, target):
  
  dic = {}

  for i in arr:
    if i > target:
      dic[i] = 0
    else:
      dic[i] = 1

  for key in dic:
    if key in dic:
      if not target-key == key:
        if target-key in dic:
          print(True)
          return

  print(False)

hash([1,2,3,4,8], 6)
hash([2,3,5,9], 10)

File: jijun0129/Algorithm/18.js
function countSort(arr, k) {
  const hashtable = new Array(k + 1).fill(0);
  for (const num of arr) {
    if (num <= k) {
      // 값을 인덱스로 가지는 hashtable을 생성한다.
      hashtable[num] = 1;
    }
  }
  return hashtable;
}

function solution(arr, target) {
  const hashtable = countSort(arr, target);
  for (const num of arr) {
    const complement = target - num; // 남은 숫자가 필요한 숫자
    if (
      complement !== num &&
      complement >= 0 &&
      complement <= target &&
      hashtable[complement] === 1 // 남은 숫자가 hashtable에 존재 -> 값이 존재한다.
    ) {
      return true;
    }
  }
  return false;
}
console.log(solution([1, 2, 3, 4, 8], 6));
console.log(solution([2, 3, 5, 9], 10));
